### DISCUSSION TITLE
Build failures on macOS Sonoma 14.1

### DISCUSSION THREAD

**Discussion by edwardloveall** [rbenv/ruby-build#2309] [Build failures] [answered]
Hello! This isn't urgent but I'm posting here to rubber duck, potentially get some community support, and potentially act as a landing point for others with similar issues.

I'm unable to install any ruby version. I'm using `ruby-build` via [asdf](https://github.com/asdf-vm/asdf-ruby) but I get the same errors when using `ruby-build` directly:

```
asdf install ruby 3.2.2
ruby-build: using openssl@3 from homebrew
==> Downloading ruby-3.2.2.tar.gz...
-> curl -q -fL -o ruby-3.2.2.tar.gz https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.2.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 19.5M  100 19.5M    0     0  14.1M      0  0:00:01  0:00:01 --:--:-- 14.1M
==> Installing ruby-3.2.2...
ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.asdf/installs/ruby/3.2.2" --with-openssl-dir=/opt/homebrew/opt/openssl@3 --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp

BUILD FAILED (macOS 14.1 on arm64 using ruby-build 20231107)

You can inspect the build directory at [...snip...]
See the full build log at [...snip...].log
```

Inspecting the build log, I see this (edited very slightly to try and highlight the important bits):

```
cd /var/folders/[...snip...]
==> Downloading ruby-3.2.2.tar.gz...
-> curl -q -fL -o ruby-3.2.2.tar.gz https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.2.tar.gz
-> tar xzf ruby-3.2.2.tar.gz
cd /var/folders/[...snip...]/ruby-3.2.2
==> Installing ruby-3.2.2...
-> ./configure "--prefix=$HOME/.asdf/installs/ruby/3.2.2" --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-openssl-dir=/opt/homebrew/opt/openssl@3
checking for ruby... /Users/edwardloveall/.asdf/shims/ruby
tool/config.guess already exists
tool/config.sub already exists
checking build system type... aarch64-apple-darwin23.1.0
checking host system type... aarch64-apple-darwin23.1.0
checking target system type... aarch64-apple-darwin23.1.0
checking for llvm-ar... llvm-ar
checking for clang++... clang++
checking for llvm-nm... llvm-nm
checking for llvm-objcopy... llvm-objcopy
checking for llvm-objdump... llvm-objdump
checking for llvm-ranlib... llvm-ranlib
checking for llvm-strip... llvm-strip
checking for gcc... clang
checking whether the C compiler works... no
configure: error: in `/var/folders/[...snip...]/ruby-3.2.2':
configure: error: C compiler cannot create executables
See `config.log' for more details
external command failed with status 77
```

The lines that stood out to me: 

```
checking whether the C compiler works... no
configure: error: in `/var/folders/[...snip...]/ruby-3.2.2':
configure: error: C compiler cannot create executables
```

I also looked in the `config.log`. It's long, but I'll paste what feel like potentially relevant bits:

```
[...snip...]

configure:5581: checking for gcc
configure:5613: result: clang
configure:5966: checking for C compiler version
configure:5975: clang --version >&5
Homebrew clang version 11.1.0
Target: arm64-apple-darwin23.1.0
Thread model: posix
InstalledDir: /opt/homebrew/bin

[...snip...]

configure:6006: checking whether the C compiler works
configure:6028: clang    conftest.c  >&5
dyld[79168]: Symbol not found: __ZNKSt3__115basic_stringbufIcNS_11char_traitsIcEENS_9allocatorIcEEE3strEv
  Referenced from: <450E2E3E-9F4C-31AD-939B-2D0C5952A0CB> /opt/homebrew/Cellar/llvm@11/11.1.0_4/lib/libLLVM.dylib
  Expected in:     <8B7A3A8B-D9C3-3439-90B9-FF6D9EBFB58E> /opt/homebrew/Cellar/llvm@11/11.1.0_4/lib/libc++.1.0.dylib
clang-11: error: unable to execute command: Abort trap: 6
clang-11: error: linker command failed due to signal (use -v to see invocation)
configure:6032: $? = 254
configure:6072: result: no

[...snip...]
```

## What I've tried

* Reinstalling openssl `brew reinstall openssl`
* Running the installer with `RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)"`
* Running the installer with `CFLAGS="-Wno-error=implicit-function-declaration"`
* Running the installer with all the flag listed in https://github.com/asdf-vm/asdf-ruby/issues/198
* Reinstalling Apple's command line tools
* Ensured the command line tools are selected: `sudo xcode-select -s /Library/Developer/CommandLineTools`
* `sudo xcode-select --reset` as mentioned in https://github.com/rbenv/ruby-build/discussions/2305
* `brew reinstall llvm`
* installed the ruby-build dependencies suggested in https://github.com/rbenv/ruby-build/discussions/2178#discussioncomment-5509177

Any other suggestions? I'm happy to try almost anything and/or produce more logs.

**Comment by edwardloveall** [comment, marked as answer]
Turns out, I had a wrong version of LLVM installed, along with the correct version.

Looking through the `config.log`, I (again) saw these lines:

```
dyld[79168]: Symbol not found: __ZNKSt3__115basic_stringbufIcNS_11char_traitsIcEENS_9allocatorIcEEE3strEv
  Referenced from: <450E2E3E-9F4C-31AD-939B-2D0C5952A0CB> /opt/homebrew/Cellar/llvm@11/11.1.0_4/lib/libLLVM.dylib
  Expected in:     <8B7A3A8B-D9C3-3439-90B9-FF6D9EBFB58E> /opt/homebrew/Cellar/llvm@11/11.1.0_4/lib/libc++.1.0.dylib
```

I noticed that `llvm@11` seemed old because I just read that homebrew [installs LLVM version 17](https://formulae.brew.sh/formula/llvm). Turns out I had several versions of llvm installed locally. 

After running `brew uninstall llvm@11` (and also uninstalling `gomuks` which for some reason depended on `llvm@11`) I was able to install ruby `3.2.2`.

**Reply by speric** [reply]
This worked for me.

**Reply by iason-saganas** [reply]
`brew install llvm@17` worked for me.

**Comment by maiquitome** [comment]
Worked for me: https://github.com/asdf-vm/asdf-ruby/issues/198 and need ```brew install openssl@1.1```

**Reply by lectricas** [reply]
this one helped me, not the marked answer.

**Comment by pablogravielseo** [comment]
@edwardloveall I have this problem, I'm using Sonoma 14.1.1 and followed all the steps you mentioned, but still have this problem. In my case I need to install Ruby 3.0.2.

**Reply by edwardloveall** [reply]
Sorry to hear that @pablogravielseo. Sounds like it may be a different issue if those solutions didn't work for you. I'd guess that your best option is to open a new issue or discussion with your error messages and what you tried.

**Reply by h3ct0rjs** [reply]
@pablogravielseo did you fix the issue ?

**Reply by pablogravielseo** [reply]
@h3ct0rjs, @edwardloveall was right, my problem is not related directly to this discussion issue, I was able to fix configuring the directory where OpenSSL is installed on my machine using version 1.1 using: 

`RUBY_CONFIGURE_OPTS="--with-openssl-dir=/opt/homebrew/opt/openssl@1.1" rbenv install 3.0.2` 

Thank you!

**Comment by osmanbowser** [comment]
I ran into the same issue.

Turns out ruby-build failed because libyaml was not installed. 
Vis: https://github.com/asdf-vm/asdf-ruby/issues/198#issue-787772964

Quick fix:
1. Install libyaml `brew install libyaml`
2. Install ruby-build `asdf install ruby 3.2.2`

**Reply by shamkeev** [reply]
thanks, this helped in my case on macos sonoma
